jib-core/src/main/java/com/google/cloud/tools/jib/ncache/DefaultCacheStorageWriter.java

```diff
@@ -154,30 +220,60 @@ CacheEntry write(CacheWrite cacheWrite) throws IOException {
               .setLayerDiffId(writtenLayer.layerDiffId)
               .setLayerSize(writtenLayer.layerSize)
               .setLayerBlob(Blobs.from(layerFile));
-      if (cacheWrite.getMetadataBlob().isPresent()) {
+      if (uncompressedCacheWrite.getMetadataBlob().isPresent()) {
         Path metadataFile = defaultCacheStorageFiles.getMetadataFile(writtenLayer.layerDigest);
         cacheEntryBuilder.setMetadataBlob(Blobs.from(metadataFile)).build();
       }
 
       // Write the selector file.
-      if (cacheWrite.getSelector().isPresent()) {
-        writeSelector(cacheWrite.getSelector().get(), writtenLayer.layerDigest);
+      if (uncompressedCacheWrite.getSelector().isPresent()) {
+        writeSelector(uncompressedCacheWrite.getSelector().get(), writtenLayer.layerDigest);
       }
 
       return cacheEntryBuilder.build();
     }
   }
 
   /**
-   * Writes the {@code layerBlob} to the {@code layerDirectory}.
+   * Writes a compressed {@code layerBlob} to the {@code layerDirectory}.
    *
-   * @param layerBlob the layer {@link Blob}
+   * @param compressedLayerBlob the compressed layer {@link Blob}
    * @param layerDirectory the directory for the layer
    * @return a {@link WrittenLayer} with the written layer information
    * @throws IOException if an I/O exception occurs
    */
-  private WrittenLayer writeLayerBlobToDirectory(Blob layerBlob, Path layerDirectory)
-      throws IOException {
+  private WrittenLayer writeCompressedLayerBlobToDirectory(
+      Blob compressedLayerBlob, Path layerDirectory) throws IOException {
+    // Writes the layer file to the temporary directory.
+    Path temporaryLayerFile = Files.createTempFile(layerDirectory, null, null);
+    temporaryLayerFile.toFile().deleteOnExit();
+
+    BlobDescriptor layerBlobDescriptor;
+    try (OutputStream fileOutputStream =
+        new BufferedOutputStream(Files.newOutputStream(temporaryLayerFile))) {
+      layerBlobDescriptor = compressedLayerBlob.writeTo(fileOutputStream);
+    }
+
+    // Gets the diff ID.
+    DescriptorDigest layerDiffId = getDiffIdByDecompressingFile(temporaryLayerFile);
+
+    Path layerFile = layerDirectory.resolve(defaultCacheStorageFiles.getLayerFilename(layerDiffId));
+    moveIfDoesNotExist(temporaryLayerFile, layerFile);
```

Comment:
It'd be nice if we remove the double moves (first moving a temp file in a temp directory here, and then moving the temp directory by the caller). However, I am fine if the double moves keep the code more readable.

Reply:
The moving the temp file is to name it correctly (with the diff ID, which can only be obtained after the file is written). The moving the temp directory is to atomically move the directory into place rather having the directory in the final location with empty contents before the layer contents are added.